### PR TITLE
solana: router admin

### PR DESCRIPTION
**Pull request by bingyuyap**
This PR adds admin related functionality for the Solana GMP Router:

For more info: check out [svm/programs/router/README.md](https://github.com/wormholelabs-xyz/poc-gmp-router/pull/8/files#diff-ec74889e6f44fc4fb23a81e601f1bd0c2e3c176d7c5bdf1d44a42670a36accdc)

The table shows the instructions, the specs and also the tests that verifies the specs. 

Do not worry, `5,559` lines out of `9,206` lines are from `Cargo.lock` and `2,729` are from tests! 

Events emission is not included in this PR yet!

### CODE REVIEW

File: svm/programs/router/src/instructions/add_transceiver.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+///
+/// Returns `Ok(())` if the transceiver is successfully registered, or an error otherwise.
+#[access_control(AddTransceiver::validate(&ctx.accounts))]
+pub fn add_transceiver(ctx: Context<AddTransceiver>, args: AddTransceiverArgs) -> Result<()> {
+    // Check if there's a pending_config
+    if ctx.accounts.integrator_config.pending_admin.is_some() {
+        return Err(AdminTransferInProgress.into());
+    }
+
+    let transceiver_id = ctx.accounts.integrator_config.registered_transceivers.len() as u8;
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/add_transceiver.rs:89]: could call this `index` and simplify line 100

File: svm/programs/router/src/instructions/add_transceiver.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    /// The integrator config account
+    /// This makes sure that the admin signing this ix is the one registed in the IntegratorConfig
+    /// The new registered transceiver will be pushed to the `registered_transceivers` field in
+    /// this account
+    /// `has_one` constraint checks if admin signer is the current admin of the config
+    #[account(
+        mut,
+        seeds = [IntegratorConfig::SEED_PREFIX, args.integrator_program_id.as_ref()],
+        bump = integrator_config.bump,
+        has_one = admin @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/add_transceiver.rs:41]: this is redundant with `check_admin`

File: svm/programs/router/src/instructions/disable_transceiver.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /// The admin account that has the authority to disable transceivers
+    pub admin: Signer<'info>,
+
+    /// The integrator config account
+    /// The account constraints here make sure that the one signing this transaction is the admin
+    /// of the config
+    #[account(
+        seeds = [IntegratorConfig::SEED_PREFIX, args.integrator_program_id.as_ref()],
+        bump = integrator_config.bump,
+        has_one = admin @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/disable_transceiver.rs:18]: this is redundant with `check_admin`

File: svm/programs/router/src/instructions/enable_transceiver.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    pub admin: Signer<'info>,
+
+    /// The integrator config account
+    /// The account constraints here make sure that the one signing this transaction is the admin
+    /// of the config
+    /// The `has_one` constraint checks if admin signer is the current admin of the config
+    #[account(
+        seeds = [IntegratorConfig::SEED_PREFIX, args.integrator_program_id.as_ref()],
+        bump = integrator_config.bump,
+        has_one = admin @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/enable_transceiver.rs:23]: this is redundant with `check_admin`

File: svm/programs/router/src/instructions/transfer_admin.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    /// The IntegratorConfig account being transferred
+    /// `has_one` constraint checks that the signer is the current admin
+    #[account(
+        mut,
+        seeds = [
+            IntegratorConfig::SEED_PREFIX,
+            args.integrator_program_id.key().as_ref(),
+        ],
+        bump = integrator_config.bump,
+        has_one = admin @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/transfer_admin.rs:29]: this is redundant with `check_admin`

File: svm/programs/router/src/instructions/update_admin.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    /// The IntegratorConfig account being transferred
+    /// `has_one` constraint checks that the signer is the current admin
+    #[account(
+        mut,
+        seeds = [
+            IntegratorConfig::SEED_PREFIX,
+            args.integrator_program_id.key().as_ref(),
+        ],
+        bump = integrator_config.bump,
+        has_one = admin @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/update_admin.rs:29]: this is redundant with `check_admin`

File: svm/programs/router/src/state/integrator_chain_config.rs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    /// Bitmap tracking the status of send transceivers
+    pub send_transceiver_bitmap: Bitmap,
+}
+
+impl IntegratorChainConfig {
+    /// Seed prefix for deriving IntegratorChainConfig PDAs
+    pub const SEED_PREFIX: &'static [u8] = b"integrator_chain_config";
+
+    pub fn new(bump: u8, integrator_program_id: Pubkey, chain_id: u16) -> Self {
```
Comment (evan-gray) [inline svm/programs/router/src/state/integrator_chain_config.rs:34]: feel like I'm missing something, where is this used?

File: svm/programs/router/src/instructions/enable_transceiver.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    )]
+    pub integrator_config: Account<'info, IntegratorConfig>,
+
+    /// The integrator chain config account
+    /// This account will be initialized if it doesn't exist, and its bitmap will be updated
+    #[account(
+        init_if_needed,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/enable_transceiver.rs:30]: how does this set the initial account data? I'm not seeing `new` called anywhere in this file, but maybe I'm missing something.

File: svm/programs/router/src/instructions/update_admin.rs [hunk 9]
```diff
@@ -0,0 +1,66 @@
+use crate::error::RouterError;
+use crate::state::IntegratorConfig;
+use anchor_lang::prelude::*;
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+pub struct UpdateAdminArgs {
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/update_admin.rs:6]: personally, I might have put all of the admin transfer/claim/update/discard ixs in the same `admin.rs` file and shared this struct, but can do that clean up in a follow up, if desired.

File: svm/programs/router/src/instructions/transfer_admin.rs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+#[derive(Accounts)]
+pub struct ClaimAdmin<'info> {
+    /// The signer, which can be either the pending_admin or the current admin
+    pub new_admin: Signer<'info>,
+
+    /// The IntegratorConfig account being claimed
+    /// The constraint here checks that the signer is either the pending_admin or the current_admin
+    #[account(
+        mut,
+        constraint = (integrator_config.pending_admin == Some(new_admin.key()) || integrator_config.admin == new_admin.key()) @ RouterError::CallerNotAuthorized,
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/transfer_admin.rs:49]: I don't think that `MUST check that there is an admin transfer pending` is covered.

File: svm/programs/router/src/instructions/transfer_admin.rs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+/// * `ctx` - The context of the request, containing the accounts involved in the admin update.
+/// * `args` - The TransferAdminArg struct containing the new admin's public key.
+///
+/// # Returns
+///
+/// Returns `Ok(())` if setting the pending admin is successful, otherwise returns an error.
+#[access_control(TransferAdmin::validate(&ctx.accounts))]
+pub fn transfer_admin(ctx: Context<TransferAdmin>, args: TransferAdminArgs) -> Result<()> {
+    // Checks that there is no pending transfer
+    if ctx.accounts.integrator_config.pending_admin.is_some() {
```
Comment (evan-gray) [inline svm/programs/router/src/instructions/transfer_admin.rs:73]: This is a duplicate of `check_admin`

File: svm/programs/router/src/instructions/transfer_admin.rs (hunk 10 above)
Comment (evan-gray) [reply svm/programs/router/src/instructions/transfer_admin.rs:49]: Also this doesn't check the immutable flag, but maybe we can get rid of that anyway.

File: svm/programs/router/src/instructions/update_admin.rs (hunk 9 above)
Comment (bingyuyap) [reply svm/programs/router/src/instructions/update_admin.rs:6]: [issue](https://github.com/wormholelabs-xyz/example-gmp-router/issues/22) created

File: svm/programs/router/src/instructions/transfer_admin.rs (hunk 10 above)
Comment (bingyuyap) [reply svm/programs/router/src/instructions/transfer_admin.rs:49]: added the constraints 
```
        constraint = integrator_config.admin.is_some() @ RouterError::CallerNotAuthorized,
        constraint = integrator_config.pending_admin.is_some() @ RouterError::NoAdminTransferInProgress,
        constraint = integrator_config.pending_admin == Some(new_admin.key()) || integrator_config.admin == Some(new_admin.key()) @ RouterError::CallerNotAuthorized,
```

I don't see EVM having `NoAdminTransferInProgress` so lmk if we want to explicitly throw if there is no pending transfer. It is implicitly checked in `integrator_config.pending_admin == Some(new_admin.key())` but not `integrator_config.admin == Some(new_admin.key())`

File: svm/programs/router/src/instructions/enable_transceiver.rs (hunk 8 above)
Comment (evan-gray) [reply svm/programs/router/src/instructions/enable_transceiver.rs:29]: The other fields of `IntegratorChainConfig` are never populated though. the `bump`, `integrator_program_id`, and `chain_id` are never set. These instructions should have some way of determining if the account is initialized or not and set those fields (or a method on `IntegratorChainConfig` which does both). In this case, you may be able to use `chain_id != 0` if you enforce in `validate` that the `chain_id` arg cannot be `0`

**Review by evan-gray — CHANGES_REQUESTED** [review]
getting close!

File: svm/programs/router/src/state/integrator_config.rs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    pub fn pda(integrator_program_id: &Pubkey) -> (Pubkey, u8) {
+        Pubkey::find_program_address(
+            &[Self::SEED_PREFIX, integrator_program_id.as_ref()],
+            &crate::ID,
+        )
+    }
+
+    pub fn check_admin(&self, signer: &Signer) -> Result<()> {
+        require!(
+            self.admin.is_some() && self.admin == Some(signer.key()),
```
[the remark this answers is not in the document]
Comment (bingyuyap) [reply svm/programs/router/src/state/integrator_config.rs:46]: image

File: svm/programs/router/src/instructions/update_admin.rs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+///
+/// Returns `Ok(())` if the admin update is successful, otherwise returns an error.
+#[access_control(UpdateAdmin::validate(&ctx.accounts))]
+pub fn update_admin(ctx: Context<UpdateAdmin>, args: UpdateAdminArgs) -> Result<()> {
+    // Check if there's a pending admin transfer
+    if ctx.accounts.integrator_config.pending_admin.is_some() {
+        return Err(RouterError::AdminTransferInProgress.into());
+    }
+
+    ctx.accounts
```
[the remark this answers is not in the document]
Comment (bingyuyap) [reply svm/programs/router/src/instructions/update_admin.rs:60]: updated

File: svm/programs/router/src/state/integrator_config.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (bingyuyap) [reply svm/programs/router/src/state/integrator_config.rs]: updated

File: svm/programs/router/src/instructions/enable_transceiver.rs (hunk 8 above)
Comment (bingyuyap) [reply svm/programs/router/src/instructions/enable_transceiver.rs:30]: updated, added a check where it fails with `InvalidChainId` when `chain_id` = 0 is passed in as arg. We might want to add this to EVM as well

Comment (evan-gray) [reply svm/programs/router/src/instructions/enable_transceiver.rs:30]: I think EVM does have this 

https://github.com/wormholelabs-xyz/example-gmp-router/blob/633b4d9ad7c00560804dce92fb88dda1088545a0/evm/src/TransceiverRegistry.sol#L189-L191

**Review by evan-gray — APPROVED** [review]

**Pull request merged**